filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+// changes in order to refresh its stored data.
+type FileWatcher[T any] interface {
+	// Close stops the FileWatcher
+	//
+	// After Close is called you won't get any updates on the file content,
+	// but you can still call Get to get the last content before stopping.
+	//
+	// It's OK to call Close multiple times.
+	// Calls after the first one are essentially no-op.
+	//
+	// Close never return an error.
+	io.Closer
+
+	// Get returns the latest, parsed data from the FileWatcher.
+	Get() T
+}
+
+// A Parser is a callback function to be called when a watched file has its
+// content changed, or is read for the first time.
+type Parser[T any] func(f io.Reader) (data T, err error)
+
+// A DirParser is a callback function that will be called when the watched
+// directory has its content changed or is read for the first time.
+//
+// Use WrapDirParser to wrap it into a Parser to be used with FileWatcher.
+type DirParser[T any] func(dir fs.FS) (data T, err error)
+
+// WrapDirParser wraps a DirParser for a directory to a Parser.
+//
+// When using FileWatcher to watch a directory instead of a single file,
+// you MUST use WrapDirParser instead of any other Parser implementations.
+func WrapDirParser[T any](dp DirParser[T]) Parser[T] {
+	return func(r io.Reader) (data T, err error) {
+		path := string(r.(fakeDirectoryReader))
+		dir := os.DirFS(path)
+		return dp(dir)
+	}
+}
+
+// Result is the return type of New. Use Get function to get the actual data.
+type Result[T any] struct {
+	data atomic.Value
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Get returns the latest parsed data from the FileWatcher.
+func (r *Result[T]) Get() T {
+	return r.data.Load().(*atomicData[T]).data
+}
+
+// Close stops the FileWatcher.
+//
+// After Close is called you won't get any updates on the file content,
+// but you can still call Get to get the last content before stopping.
+//
+// It's OK to call Close multiple times.
+// Calls after the first one are essentially no-op.
+//
+// Close never returns an error.
+func (r *Result[T]) Close() error {
+	r.cancel()
+	return nil
+}
+
+func getMtime(path string) (time.Time, error) {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return stat.ModTime(), nil
+}
+
+func (r *Result[T]) watcherLoop(
+	watcher *fsnotify.Watcher,
+	path string,
+	parser Parser[T],
+	softLimit, hardLimit int64,
+	pollingInterval time.Duration,
+	fsEventsDelay time.Duration,
+) {
+	var lock sync.Mutex
+	forceReload := func() {
+		// make sure we don't run forceReload concurrently
+		lock.Lock()
+		defer lock.Unlock()
+
+		d, mtime, files, err := openAndParse(path, parser, softLimit, hardLimit)
+		if err != nil {
+			slog.ErrorContext(r.ctx, "filewatcher: openAndParse returned error", "err", err)
+		} else {
+			r.data.Store(&atomicData[T]{
+				data:  d,
+				mtime: mtime,
+			})
+			// remove all previously watched files
+			for _, path := range watcher.WatchList() {
+				watcher.Remove(path)
```

Comment:
Do we care about logging the error here at all? Though I feel like it'll make it just unnecessarily noisy.